staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -758,13 +774,20 @@ func (s *store) ttlOpts(ctx context.Context, ttl int64) ([]clientv3.OpOption, er
 	if ttl == 0 {
 		return nil, nil
 	}
-	// TODO: one lease per ttl key is expensive. Based on current use case, we can have a long window to
-	// put keys within into same lease. We shall benchmark this and optimize the performance.
-	lcr, err := s.client.Lease.Grant(ctx, ttl)
-	if err != nil {
-		return nil, err
+	s.leaseLock.Lock()
+	defer s.leaseLock.Unlock()
+	// If the previous lease grants valid TTL, reuse it; Otherwise create
+	// a new one with a little extra time to make it possible to be reused.
+	now := time.Now().Unix()
+	if now+ttl > s.prevLeaseExpirationTime {
```

Comment:
this approach (monotonically increasing) only works if all ttls are identical. it might make more sense to bucket ttls around the same time, and create one lease per bucket (can control the granularity of the bucket, and ensure the lease for the bucket expires at the max of all the ttls in the bucket)

Reply:
The failure mode is some item with a realllllly long lease comes along and then holds many subsequent things open, right?

I don't think we actually do anything that would exercise that case.

Keeping a few buckets around is better but more complex--I don't know that we need it.

Reply:
> we actually do anything that would exercise that case

kube-apiserver doesn't, but other consumers of `k8s.io/apiserver` do

Reply:
I am not familiar with the use case. However, TMO, the following would work.
```
if ttl > s.prevLeaseExpirationTime-now || ttl*1.5 < s.prevLeaseExpirationTime-now {
```
It avoids the complex buckets, while the failing case would be events with large and small TTLs interleave with each other one by one.

The magic number 1.5 can be changed to any number based on our use cases, e.g., if we expect the events to expire in 1.5 * TTL.